File: 7.Stringspermutations/C/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

void swap(char* a, char* b) {
    char temp = *a;
    *a = *b;
    *b = temp;
}

void permute(char* str, int l, int r, int* count) {
    if (l == r) {
        (*count)++;
    } else {
        for (int i = l; i <= r; i++) {
            swap((str + l), (str + i));
            permute(str, l + 1, r, count);
            swap((str + l), (str + i));
        }
    }
}

void test(const char* str)
{
    int len = strlen(str);
    char* str_mut = malloc(sizeof(char) * len);
    strcpy(str_mut, str);
    printf("String to permute: %s, len: %d\n", str, len);
    int count = 0;

    clock_t start_time = clock();
    permute(str_mut, 0, len - 1, &count);
    clock_t end_time = clock();

    double execution_time = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    printf("Permutations: %d\n", count);
    printf("Execution Time: %.5f seconds\n\n", execution_time);
}

int main() {
    test("abcdefghij");
	test("abcdefghijk");
	test("abcdefghijkl");

    return 0;
}

File: 7.Stringspermutations/Go/main.go
package main

import (
	"fmt"
	"time"
)

func swap(str []rune, i, j int) {
	str[i], str[j] = str[j], str[i]
}

func permute(str []rune, l, r int, count *int) {
	if l == r {
		(*count)++
	} else {
		for i := l; i <= r; i++ {
			swap(str, l, i)
			permute(str, l+1, r, count)
			swap(str, l, i)
		}
	}
}

func test(str string) {
	fmt.Printf("String to permute: %s, len: %d\n", str, len(str))
	runes := []rune(str)
	count := 0

	startTime := time.Now()
	permute(runes, 0, len(runes)-1, &count)
	endTime := time.Now()

	executionTime := endTime.Sub(startTime)

	fmt.Println("Permutations:", count)
	fmt.Printf("Execution time: %.6f seconds\n\n", executionTime.Seconds())
}

func main() {
	test("abcdefghij")
	test("abcdefghijk")
	test("abcdefghijkl")
}
